File: atividade-20/atividade.py
class AFD:
    def __init__(self):
        self.state = 'q0'
    
    def transition(self, char):
        if self.state == 'q0':
            if char == 'a':
                self.state = 'q1'
        elif self.state == 'q1':
            if char == 'b':
                self.state = 'q2'
            elif char == 'a':
                self.state = 'q1'
        elif self.state == 'q2':
            if char == 'a':
                self.state = 'q3'
        elif self.state == 'q3':
            pass  
    
    def is_accepted(self, string):
        self.state = 'q0'  
        for char in string:
            self.transition(char)
        return self.state == 'q2'

if __name__ == '__main__':
    afd = AFD()
    print(afd.is_accepted("ab"))      
    print(afd.is_accepted("aab"))     
    print(afd.is_accepted("abab"))    
    print(afd.is_accepted("baab"))    
    print(afd.is_accepted("bbaab"))   
    print(afd.is_accepted("baabab"))  
    print(afd.is_accepted("aaaaa"))  
    print(afd.is_accepted("bbbbb"))  

File: atividade-20/atividade.go
package main

import "fmt"

type AFD struct {
	state string
}

func NewAFD() *AFD {
	return &AFD{state: "q0"}
}

func (afd *AFD) transition(c rune) {
	switch afd.state {
	case "q0":
		if c == 'a' {
			afd.state = "q1"
		}
	case "q1":
		if c == 'b' {
			afd.state = "q2"
		} else if c == 'a' {
			afd.state = "q1"
		}
	case "q2":
		if c == 'a' {
			afd.state = "q3"
		}
	case "q3":
		
	}
}

func (afd *AFD) isAccepted(input string) bool {
	afd.state = "q0"
	for _, c := range input {
		afd.transition(c)
	}
	return afd.state == "q2"
}

func main() {
	afd := NewAFD()
	fmt.Println(afd.isAccepted("ab"))     
	fmt.Println(afd.isAccepted("aab"))    
	fmt.Println(afd.isAccepted("abab"))   
	fmt.Println(afd.isAccepted("baab"))   
	fmt.Println(afd.isAccepted("bbaab"))  
	fmt.Println(afd.isAccepted("baabab")) 
}
